serveradmin/api/api.py

```diff
@@ -0,0 +1,37 @@
+import hashlib
+from datetime import timedelta
+
+from django.utils import timezone
+
+from serveradmin.api.decorators import api_function
+from serveradmin.api.models import Lock
+
+
+@api_function(group='api')
+def lock(identifier, seconds=60):
+    """Mark identifier as in-use for n seconds
+
+    Allow clients to mark resources with an identifier as in-use so that
+    distributed clients can check whatever something is in use or not.
+
+    :param identifier: A unique identifier (e.g. 10.0.0.1)
+    :param seconds: seconds until the lock expires
+
+    :return: True on success or seconds left if already in use
+    """
+
+    # Use hash sum because this has a constant length
+    hash_sum = hashlib.sha1(str(identifier).encode()).hexdigest()
+
+    # Remove expired logs first
+    Lock.objects.filter(until__lt=timezone.now()).delete()
+
+    in_use = Lock.objects.filter(hashsum=hash_sum).only('until')
+    if in_use.exists():
+        return (in_use.get().until - timezone.now()).seconds
+
+    until = timezone.now() + timedelta(seconds=seconds)
+    Lock.objects.create(hashsum=hash_sum, duration=seconds, until=until)
```

Comment:
This should most likely be in a locking transaction, so the reading of the existence and the actual getting resp. the creation of a new one is protected for Race Condition.

Reply:
The `select_for_update` should then be used, too.
https://docs.djangoproject.com/en/3.1/ref/models/querysets/#select-for-update

Reply:
Whereby, the Race Condition is not a Security risk here actually as follow up queries would result into db errors. Still might be interesting for UX purposes.

Reply:
True, might be worth it here.

Reply:
`select_for_update` would not have been worked because the objects did not exist before. However I found a way to make the creation of the object atomic, this should do the trick.